### DISCUSSION TITLE
“Advanced Audio Settings” sub-menu

### DISCUSSION THREAD

**Discussion by Jarran2R** [ppy/osu#27058] [Ideas]
My idea is that there should be a button right under the audio output device drop-down named “Advanced Audio Settings”, and when you click it it’ll open up a sub-menu in the same way that the keybindings button does. Inside this sub-menu would be the following:
> Audio subsystem (drop-down)

Windows would have WASAPI/DirectSound as options, Linux would have ALSA/PulseAudio/JACK/PipeWire as options, and Mac only has CoreAudio if i recall correctly so maybe this drop-down could be disabled or disappear for OSes that don’t have multiple audio backends
> Audio buffer size (slider)

This would change the value of the buffer length in BASS. This is something that has the potential to cause a lot of issues, which is why i think it’s appropriate to have it in an “advanced” audio menu. I don’t like how right now it’s hardcoded to 10ms with no way to officially change it for those that understand what it does, so for this you could have the default at 10ms so that things don’t break, but then the user could change it if they wanted to. (and it’s hard to accidentally change since it’s in it’s own thing)
> Master/Music/Effect audio device (drop-downs)

These drop-downs would let you set different audio devices for music and effects. I feel like would have huge potential in the realm of streaming and content creation as you could route all the potentially copyrighted music to a separate audio device where you could remove it from VODs or edit it out of a video, and still keep the hitsounds and menu sounds. The master audio device is the device all sound goes in. Defaults would be that all three are set to the output device option in the main settings. Anything set to the same device as the master device is ignored, for example if music and master are both set to device A and effect is set to device B, device A would have master sound and device B would have effects. Either prevent music and effect devices from being the same thing, or just have it combine them and end up being the same as master.

One quick sub-idea: adding WASAPI exclusive mode and ASIO support for windows for ultra low-latency, being options under the “Audio subsystem” setting.

I’ve had these ideas for a while now and never thought to make a discussion of them until now. Thanks for reading.

**Comment by sineplusx** [comment]
I like your ideas in general. However, I have some comments:

> Audio subsystem (drop-down)

BASS on Linux only supports ALSA. Unless there is a [change away from BASS in the future](https://github.com/ppy/osu-framework/pull/6002), this dropdown would probably be exclusive to Windows. I dont know if Android / iOS have multiple backends, but I dont think so.


> Audio buffer size (slider)

This has been tried before, see #10892 and the related framework PR. You already pointed out the reason why this is going to be hard to implement: "This is something that has the potential to cause a lot of issues". Most players dont know what this option actually does behind the scenes except for "lower number = better!!!". No "advanced" menu or warning screens will stop somebody clueless from dialing this to the lowest value and then complaining about audio crackling, similarly to how people used to run unlimited frames and then complained about stutters. I think the main priority of the dev team is to not expose a setting to a user that is capable of ruining the experience, while finding defaults that work for all platforms and systems which are still reasonably low in latency.

Quick fun fact: The current default buffer size (on linux?) is not 10ms, its actually higher. This is because the period size is never explicitly set, causing BASS to default it to 10ms (441 samples). The buffer size has to be at least twice the period size, which means the buffer size ends up at 20ms (882 samples)

**Reply by bdach** [reply]
> I think the main priority of the dev team is to not expose a setting to a user that is capable of ruining the experience, while finding defaults that work for all platforms and systems which are still reasonably low in latency

Correct.

**Reply by Rekunan** [reply]
Adding on, I think it'd be much better to just have the buffer size as an environment variable.
